kubeadm: enable the v1beta4 API

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

apply a set of changes required to enable the new v1beta4 kubeadm API.
please review one commit at a time.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/2890

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: enabled the v1beta4 API. For a complete changelog since v1beta3 please see https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta4/. 

The API does include a few breaking changes:
- The "extraArgs" component construct is now a list of "name"/"value" pairs instead of a string/string map. This has been done to support duplicate args where needed.
- The "JoinConfiguration.discovery.timeout" field has been replaced by "JoinConfiguration.timeouts.discovery".
- The "ClusterConfiguration.timeoutForControlPlane" field has been replaced by "{Init|Join}Configuration.timeouts.controlPlaneComponentHealthCheck".
Please use the command "kubeadm config migrate" to migrate your existing v1beta3 configuration to v1beta4.

v1beta3 is now marked as deprecated but will continue to be supported until version 1.34 or later.
The storage configuration in the kube-system/kubeadm-config ConfigMap is now a v1beta4 ClusterConfiguration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: update godoc and comments for v1beta4

- Include some more examples related to v1beta4 in the doc.go.
- Fix some typos in v1beta4 field comments.
- Add missing JSON tag for UpgradeConfiguration.Apply.SkipPhases.

neolit123 committed:
kubeadm: add missing "kubeadm config print upgrade-defaults"

neolit123 committed:
kubeadm: move v1beta4 to high priority in API scheme

neolit123 committed:
kubeadm: update migration logic from v1beta3 to 4

Treat v1beta4 as non experimental API when migrating.
Update unit tests.

neolit123 committed:
kubeadm: don't default ExtraEnvs to an empty slice

Instead of defaulting ExtraEnvs for CP components to an empty
slice when converting from/to v1beta3 keep it nil.

This allows for expecting a nil value in the internal
config, similarly to ExtraArgs.

neolit123 committed:
kubeadm: keep v1beta3 only in required code paths

Use v1beta4 everywhere and only use v1beta3
in a few required locations:
- kubeadm-config map handling
- unit and integration tests

neolit123 committed:
kubeadm: more validation for Upgrade|ResetConfiguration

- Add unit tests for ValidateUpgrade|ResetConfiguration
- Add two more validation points in ValidateUpgradeConfiguration

neolit123 committed:
kubeadm: continue supporting extraArgs flags for v1beta4

Flags for kubeadm init such as --apiserver-extra-args prior
to v1beta4 used a map[string]string for pflag.Value storage. This no
longer works since v1beta4 extra args are a slice of Arg.

Add a new flag type argSlice and implement a solution for
parsing these flags.

At the same time deprecate these flags and show a warning
that users should use config.

neolit123 commented:
/kind feature
/priority important-soon
/triage accepted

neolit123 commented:
the needs-rebase label seems stuck.

/approve cancel
/hold

PTAL when possible, code freeze is due July first.
https://github.com/kubernetes/sig-release/tree/master/releases/release-1.31

/cc @SataQiu @pacoxu @carlory

neolit123 commented:
open topics:

- should we mark v1beta3 as deprecated in this PR?
https://kubernetes.io/docs/reference/using-api/deprecation-policy/
historically this is how we have done it...new API means deprecate the old one.

- should we write both the v1beta3 and v1beta4 clusterconfiguration in the kubeadm-config configmap for one release?
see the commit named `kubeadm: write both v1beta3 and v1beta4 in kubeadn-config`. i saw one complain from that on the side of cluster-api where someone had a use case to use kubeadm with version n-1 to join a node to a kubeadm N created cluster. this is not supported by our skew which is - a kubeadm cluster created or upgrade with kubeadm vertsion N can only allow join of nodes with kubeadm version N (not N-1 or N+1).
the change will also require cleanup on the next `kubeadm upgrade apply` maybe in 1.32 to have only one CM key.
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm

neolit123 committed:
kubeadm: deprecate v1beta3

Mark v1beta3 as deprecated by showing a warning when used
and write a note about it in the doc.go file.

neolit123 commented:
> should we mark v1beta3 as deprecated in this PR?
https://kubernetes.io/docs/reference/using-api/deprecation-policy/
historically this is how we have done it...new API means deprecate the old one.
it would need at least 3 releases / 1 year before removal.

pushed a commit for that and updated the release note.

> should we write both the v1beta3 and v1beta4 clusterconfiguration in the kubeadm-config configmap for one release?
see the commit named kubeadm: write both v1beta3 and v1beta4 in kubeadn-config. i saw one complain from that on the side of cluster-api some time ago where someone had a use case to use kubeadm with version n-1 to join a node to a kubeadm N created cluster. this is not supported by our skew which is - a kubeadm cluster created or upgrade with kubeadm vertsion N can only allow join of nodes with kubeadm version N (not N-1 or N+1). i think recent cluster-api versions actually block this.
..and the change will also require cleanup on the next kubeadm upgrade apply maybe in 1.32 to have only one CM key. maybe i can remove this commit entirely.
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm

this allows projects that do the roll-out / node-replace upgrade and manage their own logic instead of `kubeadm upgrade apply` to continue using v1beta3 for longer.

fabriziopandini commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -77,24 +78,37 @@ func getInitConfigurationFromCluster(kubeconfigDir string, client clientset.Inte
+			return nil, errors.Errorf("unexpected error when reading kubeadm-config ConfigMap: %s or %s key value pair missing",
+				constants.ClusterConfigurationConfigMapKey, constants.ClusterConfigurationNewConfigMapKey)
+		}
+		gvkToVerify = kubeadmapiv1old.SchemeGroupVersion.WithKind(constants.ClusterConfigurationKind)
```
> kubeadm 1.31 will write two keys in the kube-system/kubeadm-config ConfigMap during "init" and "upgrade apply" - "ClusterConfigurationNew" (for v1beta4) and "ClusterConfiguration" (for v1beta3).
> This change is done to ease transition for clusters that join N-1 kubeadm version nodes, however it will be restored to a single "ClusterConfiguration" (v1beta4) key in a follow up release.

I think we should be really careful in declaring that we allow to use kubeadm N-1 to be used to join:
- this requires proper test coverage, and I don't know if it worth the effort to do so if this is just temporarily (and we know this cannot work in future, depending by what we change in a release)
- the two API versions cannot be converted cleanly (you might lose Extra args or timeouts), and thus even if we store two versions of the kubeadm config, we cannot guarantee they are the same. 

With regard to having two config maps, If we want to do so, I think we should find a way that do not force the users willing to opt in to v1beta1 immediately, to avoid to use the new config map temporarily.

E.g. 

- Init should always create only one config map, in v1beta4 format (green field)
- During kubeadm upgrade, if there is a v1beta3, keep it and create the new config map in v1beta4 (brown field)
- Join should be able to deal with both green and brown field, using v1beta4 config no matter if stored in ClusterConfiguration or ClusterConfigurationNew

neolit123 replied:
> I think we should be really careful in declaring that we allow to use kubeadm N-1 to be used to join:

i think the current state of the kubeadm vs kubeadm skew is not ideal, so this can be an experiment if one day this skew is to be changed. i think it should be N vs N-1, N, N+1 instead of N vs N.

> this requires proper test coverage, and I don't know if it worth the effort to do so if this is just temporarily (and we know this cannot work in future, depending by what we change in a release)

an e2e can be added once kinder is updated to support v1beta4.
it could do the following
- call "init" which creates two CM keys - v1beta3/v1beta4
- remove the v1beta4 key
- call "join" and make sure it works

> the two API versions cannot be converted cleanly (you might lose Extra args or timeouts), and thus even if we store two versions of the kubeadm config, we cannot guarantee they are the same.

but the point is that 1.30 kubeadm will be able to still join this cluster with two configmap keys created by 1.31.

> With regard to having two config maps, If we want to do so, I think we should find a way that do not force the users willing to opt in to v1beta1 immediately, to avoid to use the new config map temporarily.

> E.g.

> - Init should always create only one config map, in v1beta4 format (green field)
> - During kubeadm upgrade, if there is a v1beta3, keep it and create the new config map in v1beta4 (brown field)
> - Join should be able to deal with both green and brown field, using v1beta4 config no matter if stored in ClusterConfiguration or ClusterConfigurationNew

this cuts of 1.30 "kubeadm join", as it would not be able to decode v1beta4 which is the only key written by "init".

i don't see a way without adding the two keys. it would be a problem for users that want to use 1.31 and manually edit temporary the "new" v1beta4 key and then switch back to editing the regular CM key.

neolit123 commented:
after some discussion with @fabriziopandini on a zoom call, we decided that it's best to just not do the ClusterConfiguration, ClusterConfigurationNew separate keys and try to support the N-1 join skew. this skew support needs a better plan. as is, having the separate keys will create temporary complexity for users and also for CAPI that is not idea.

i will update the PR to remove this part.

neolit123 committed:
kubeadM: allow conversion of TimeoutForControlPlane

v1beta3.ClusterConfiguration.APIServer.TimeoutForControlPlane
must be migrated to {Init|Join}Configuration.Timeouts.
.ControlPlaneComponentHealthCheck.

To achieve this sort of cross-Kind migration do the following:
- Use a temporary, thread-safe variable in timeoututils.go
- Make the order of GVKs in documentMapToInitConfiguration
deterministic.

neolit123 commented:
added another commit that allows migration of ClusterConfiguration.timeoutForControlPlane (even if this might not have many users). it requires a hack with a temp thread safe variable and some determinism when decoding the "init" YAML multi-doc.

pacoxu commented:
/lgtm

I went back to https://github.com/kubernetes/kubernetes/pull/107013 and https://github.com/kubernetes/kubernetes/pull/101129. 

We may merge this earlier to do more test on the v1beta4 in this release cycle. A test plan/action items list may be needed then. We need to know if those new features are covered by e2e or UT.

pacoxu commented:
/approve
@SataQiu @carlory do you have time to take a look?

liggitt commented:
it seems like the 1.30 regression around timeout should be fixed in master and backported before this merges (xref https://github.com/kubernetes/kubernetes/pull/125428#issuecomment-2163494342)

neolit123 commented:
> it seems like the 1.30 regression around timeout should be fixed in master and backported before this merges (xref [#125428 (comment)](https://github.com/kubernetes/kubernetes/pull/125428#issuecomment-2163494342))

commented.
https://github.com/kubernetes/kubernetes/pull/125428#issuecomment-2163636838
i will rebase this PR after https://github.com/kubernetes/kubernetes/pull/125428 merges.

neolit123 commented:
rebased after
https://github.com/kubernetes/kubernetes/pull/125480

pacoxu commented:
/lgtm

SataQiu reviewed (commented):
Great job!
/lgtm
+1 for merging this first and doing e2e tests as much as possible

sbueringer commented:
We have a PR in Cluster API that we should be able to merge immediately after this one. CAPI doesn't use eg kubeadm upgrade, but can provide at least some extra coverage for init/join

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -24,11 +24,11 @@ limitations under the License.
-//   - `dry-run` mode in is now configurable in InitConfiguration and JoinConfiguration config files.
+//   - The ResetConfiguration API type is now supported in v1beta4. Users are able to reset a node by passing
+//     a --config file to "kubeadm reset".
+//   - `dry-run` mode in is now configurable in InitConfiguration and JoinConfiguration..
```
updated.

neolit123 replied:
also fixed typo in comment `// Te resulting ...`

over:

```
// convertToArgs takes a argument map and converts it to a slice of arguments.
// The resulting argument slice is sorted alpha-numerically.
func convertToArgs(in map[string]string) []kubeadm.Arg {
```

SataQiu commented:
/lgtm

neolit123 commented:
i will work on https://github.com/kubernetes/kubeadm/pull/3067 ASAP for e2e tests for the new API.
we have time until freeze (~10 july) for bug fixes.

neolit123 commented:
/hold cancel

neolit123 commented:
> nits in release note: `replaced by "Init|Join}` -> `replaced by "{Init|Join}`

updated.
